contract/src/withdraw.rs

```diff
@@ -102,32 +124,74 @@ impl AuroraLaunchpadContract {
+            Ok(value) => self.return_part_of_deposit(&account, value),
         }
     }
 
-    fn do_withdraw(&self, amount: U128, msg: String) -> Promise {
-        match &self.config.deposit_token {
-            DepositToken::Nep141(account_id) => ext_ft::ext(account_id.clone())
-                .with_attached_deposit(ONE_YOCTO)
-                .with_static_gas(GAS_FOR_FT_TRANSFER_CALL)
-                .ft_transfer_call(self.config.intents_account_id.clone(), amount, msg, None),
-            DepositToken::Nep245((account_id, token_id)) => ext_mt::ext(account_id.clone())
-                .with_attached_deposit(ONE_YOCTO)
-                .with_static_gas(GAS_FOR_FT_TRANSFER_CALL)
-                .mt_transfer_call(
-                    self.config.intents_account_id.clone(),
-                    token_id.clone(),
-                    amount,
-                    None,
-                    None,
-                    msg,
-                ),
+    #[private]
+    pub fn finish_mt_withdraw(
+        &mut self,
+        account: IntentsAccount,
+        amount: U128,
+        before_withdraw: (InvestmentAmount, u128, u128),
+        #[callback_result] result: &Result<Vec<U128>, PromiseError>,
+    ) {
+        require!(
+            env::promise_results_count() == 1,
+            "Expected one promise result"
+        );
+
+        // Remove the lock on the withdrawal.
+        self.locked_withdraw.remove(&account);
+
+        match result.as_deref() {
+            Ok(&[value]) if value == amount => {}
+            Ok(&[U128(0)]) | Err(_) => self.rollback_investments(account, before_withdraw),
+            Ok(&[value]) => self.return_part_of_deposit(&account, &value),
+            Ok(_) => env::panic_str("Unexpected amount of tokens withdrawn"),
         }
     }
+
+    pub fn is_withdrawal_allowed(&self, is_intents_present: bool) -> bool {
+        let status = self.get_status();
+        let is_price_discovery_ongoing = matches!(self.config.mechanics, Mechanics::PriceDiscovery)
+            && matches!(status, LaunchpadStatus::Ongoing);
+
+        (is_price_discovery_ongoing && is_intents_present)
+            || matches!(status, LaunchpadStatus::Failed)
+            || matches!(status, LaunchpadStatus::Locked)
+    }
+
+    fn rollback_investments(
+        &mut self,
+        account: IntentsAccount,
+        before_withdraw: (InvestmentAmount, u128, u128),
+    ) {
+        let (investment, total_deposited, total_sold_tokens) = before_withdraw;
+
+        self.investments.insert(account, investment);
+        self.total_deposited = total_deposited;
+        self.total_sold_tokens = total_sold_tokens;
+    }
+
+    fn return_part_of_deposit(&mut self, account: &IntentsAccount, amount: &U128) {
+        let Some(investment) = self.investments.get_mut(account) else {
+            env::panic_str("No deposits were found for the intent account");
+        };
+
+        let refund = mechanics::deposit::deposit(
```

Comment:
One potential difference between this partial refund route and the normal withdraw logic is that withdraw prevents users from getting a higher discount whereas the deposit logic will simply apply any new discount. I think this is fine because we do not expect users to be able to intentionally trigger this code path (it relies on the logic of the token, which users do not control) and so it should not be exploitable.